### DISCUSSION TITLE
❓ [QUESTION] Point cloud estimation

### DISCUSSION THREAD

**Discussion by OniDaito** [e3nn/e3nn#290] [Q&A]
Hello there,
I'm wondering if e3nn is the right tool to estimate a single point cloud from a selection of noisy point clouds? 
I've made several attempts with other methods but modelling rotation is quite difficult and my models have trouble learning.
I notice that e3nn works well on graphs, but could it also be applied to point clouds with no edges? I'm not very familiar with the mathematics employed but I do see the need for modelling rotation in a way that makes comparisons between clouds easier.
Kind regards
Ben

**Comment by mariogeiger** [comment]
I honestly don't know if e3nn is the tool you need.
It would be useful if you could describe better what you want to do, do you have examples or images to show?

Assuming you want to use a neural network to solve your problem.

1) you need a connected graph to propagate the signal
2) to create a graph from a point cloud (in an equivariant way) I would create an edge between any pair of nodes that are at a relative distance smaller than some max_radius
3) As far as I know comparing two graphs (random node indexing and rotation of the positions) is a hard task

**Comment by OniDaito** [comment]
Hello there,
Thanks for getting back to me so quickly.
The problem I want to solve would roughly be:

There are many point clouds of an underlying single object. These clouds are noisy, have missing parts but crucially, are in different orientations. Translation is easily solved / ignored by using a classic convolutional network, but rotation is much trickier.

Is it possible to define a function that can create a pointcloud that minimises the RMSD over nearest neighbours across the entire set? Given a batch of noisy pointclouds, could one generate a good average using e3nn's ability to effectively ignore the differing orientations. 

I suspect the answer might be no, as we have no edges - no graphs effectively, although one could create one by using the method you mention - linking points that are close-by.

Put another way, could e3nn 'align' a bunch of point clouds?

Cheers
Ben

**Comment by mariogeiger** [comment]
I never did it, I don't know anybody who've done it. It looks like a very important and non trivial problem.
I think that e3nn can provide useful tools for this task but it's far from obvious how I would attack this problem.

I have the feeling that the RMSD over the nearest neighbours being
- non smooth
- very local

would contain a lot of local minima.

Here is one idea that come to my mind that could avoid using the RMSD over the point clouds:
A potential approach could be to generate an equivariant feature maps of the point cloud:
```
phi: point coulds -> irreps

phi(P) = global_pool(network(P))
```

1) computes the features `phi(P1), phi(P2), phi(P3), ...`
2) align the features, i.e. find rotations `R1, R2, R3, ...` such that `Ri phi(Pi) ~= Rj phi(Pj)`

One still need to find a way to optimize `phi` such that it's represent well the point cloud.

**Comment by OniDaito** [comment]
Thank-you for helping with this.

After some more reading around I think I understand a little more what e3nn is trying to do and I suspect it's not really appropriate for my problem, however the approach you mention has been echoed in a couple of other papers I've found. 

I'll think a little more about what you've said. Thanks!

**Comment by OniDaito** [comment]
A thought occurs regarding a possibly phi, if I've understood correctly. One could represent our pointclouds by raymarching from a the surface of a sphere to the sphere origin, representing this function with a spherical harmonic. The surface of said sphere would resemble a heatmap with the intensity at a point representing the number of points passed through on the way to the origin. Something akin to this - https://github.com/daniilidis-group/spherical-cnn

**Comment by Linux-cpp-lisp** [comment]
If I'm not misinterpreting you, I think this problem is known as "point set registration": https://en.wikipedia.org/wiki/Point_set_registration

Just for future reference of people who come across this issue.

**Comment by OniDaito** [comment]
Yes, that is one way of looking at it definitely.

I think I understand a little more about the spherical harmonics and the spherical convolutional layers having read around the topic. Converting my point cloud to a spherical function to spherical harmonics and back again would be the first step for me I think. Thanks again!

**Comment by blondegeek** [comment]
Just chiming in real quick...

The fact that our convolutional filters are comprised of spherical
harmonics is simply because that is the constraint needed to preserve
equivariance. I don’t typically think of our networks first converting
things into spherical harmonics because this local projection is done at
every layer.

You could go about projecting your point cloud onto spherical harmonics — I
believe that this is a standard technique in object recognition — but you
will need a relatively high maximum L (angular frequency) and this is very
different from how I would approach the problem with e3nn.

If you would typically use CNNs for this problem then e3nns are a better
option for you use case because it simply allows you to do convolutions on
more diverse structures (not just boxes grids) and have rotation
equivariance.

The issue of how to articulate a point cloud as a graph is indeed a matter
of choice (do you use a radial cutoff or some topological method - aka
representing continuity rather than simply radial distance). But unless you
are operating on the point cloud as a whole (which you can also do with
e3nn but likely would take too much memory and be relatively inefficient)
then you are presumably making some assumption of locality (only operating
on neighborhoods at a time).

As for how to align multiple point cloud snapshots at a time — I gave some
questions for you:
+ are the snapshots inherently ordered
+ are the number of snapshots fixed

One thing that is neat to do with e3nn is to have the points “move
themselves” by translation or a global rotation. You could articulate a
task where subsequent snapshots align themselves to an initial one (note
you always need to specify an original coordinate stance to articulate a
relative rotation in 3D).

Anywho, those are some quick thoughts — please don’t hesitate to reach out
with more comments / questions.

Tess

**Comment by OniDaito** [comment]
Hello there,
Thank-you for the reply. I've been reading around the subject and I think in my case, there's a half-way house where I map my pointclouds to a spherical function and using spherical convolutions to get around the problems I'm having with rotations. 

Representing the pointcloud using some sort of minimum spanning tree or the like could be an option. I'll hold on to that.

Sadly the clouds are neither ordered, nor fixed. It's a very unrestricted problem. 

I suspect this means e3nn might not be where I start, though I do thing using convolutions and therefore hunting for local features might not be the best option  - something more global would be better. So I reckon I'll keep these ideas in mind, try some spherical convolutions first, and come back to graphs.

Cheers
Ben
